README.md
# Julia Art "juliart"

[![PyPI version](https://badge.fury.io/py/juliart.svg)](https://pypi.org/project/juliart/)
[![GitHub actions status](https://github.com/vsoch/juliart/workflows/ci/badge.svg?branch=master)](https://github.com/vsoch/juliart/actions?query=branch%3Amaster+workflow%3Aci)

This module is based on the [Color Julia Bot](https://twitter.com/colorjulia_bot)
but intended for command line generation, and with options to customize colors
based on themes (holidays, seasons) and also generate animations. 
The name "juliart" is also a play on word (Juliard) to hint
at artsy things.

![img/juliart.png](https://raw.githubusercontent.com/vsoch/juliart/master/img/juliart.png)

I've done my best so that the default parameters produce (what I think to be)
the most random and pretty images. You can of course tweak the parameters
to your liking, discussed below!

## Examples and Documentation

If you want to jump in to examples, see these repositories:

 - [vsoch/juliart-says](https://github.com/vsoch/juliart-says): write Markov generated (or raw) text onto a Julia Set image
 - [vsoch/juliart-grid](https://github.com/vsoch/juliart-grid): a grid of Julia Sets to show indexing into the Mandelbrot Set
 - [vsoch.github.io](https://vsoch.github.io/2019/juliart/): A writeup of Juliart, explaining Julia and Mandelbrot sets.


## Usage

### Install

You can install from pypi

```bash
pip install juliart
```

or install from the repository directly:

```bash
$ git clone https://github.com/vsoch/juliart
$ python setup.py install
```

for animations you'll also need an extra package:

```bash
pip install juliart[animate]
```

### Generate Art

You can see basic usage by typing `juliart` into your terminal:

```bash
$ juliart
usage: juliart [-h] [--version] {generate,animate} ...

Julia Set art generator

optional arguments:
  -h, --help          show this help message and exit
  --version           suppress additional output.

actions:
  actions for Julia Set art generator

  {generate,animate}  juliart actions
    generate          generate a Julia Set image
    animate           create a Julia Set animation (gif)
```

You'll see there are two options: to generate, and to animate. 


#### Generate

Customizations for the generate command generally include color or theme choices,
an optional text to print on the graphic, and parameters.


```bash
$ juliart generate --help
usage: juliart generate [-h] [--radius RADIUS] [--outfile OUTFILE]
                        [--text TEXT] [--fontsize FONTSIZE] [--xcoord XCOORD]
                        [--ycoord YCOORD] [--ca CA] [--cb CB] [--res RES]
                        [--iter ITERS] [--color {random,pattern,glow}]
                        [--rgb RGB]
                        [--theme {christmas,easter,fall,random,halloween,hanukkah,spring,summer,thanksgiving,valentine,winter}]
                        [--zoom ZOOM]

optional arguments:
  -h, --help            show this help message and exit
  --radius RADIUS       the max radius to allow (default is 4)
  --outfile OUTFILE     the output file to save the image (defaults to
                        randomly generated png)
  --text TEXT           write a string of text to bottom left corner
  --fontsize FONTSIZE   font size of text (if desired) defaults to 16
  --xcoord XCOORD       x coordinate for text (defaults to 0)
  --ycoord YCOORD       y coordinate for text (defaults to 0)
  --ca CA               the a component of the c parameter
  --cb CB               the b component of the c parameter
  --res RES             the resolution to generate (defaults to 1000)
  --iter ITERS          the number of iterations per pixel (defaults to 200)
  --color {random,pattern,glow}
                        a color pattern to follow.
  --rgb RGB             a specific rbg color, in format R,G,B
  --theme {christmas,easter,fall,random,halloween,hanukkah,spring,summer,thanksgiving,valentine,winter}
                        a theme to color the art (defaults to random colors)
  --zoom ZOOM           the level of zoom (defaults to 1.8)
```

##### Defaults

If you use the defaults, it will generate a randomly named image in your
present working directory.

```bash
$ juliart generate
Generating Julia Set...
Saving image to doopy-kerfuffle-5780.png
```

##### C Parameters

The values of ca and cb (the real and complex components of the c parameter
in the Julia Set equation) are chosen at random, but you can select them
to better understand how your choices influence the graphic generation:

```bash
juliart generate --ca 0.5 --cb 0.5
```

I also created a [small project](https://github.com/vsoch/juliart-grid/) that will visually show how your parameter selection
influences the resulting image! You might want to give the output file a more meaningful name:

```bash
juliart generate --ca 0.5 --cb 0.5 --outfile ca-0.5-cb-0.5.png
```

And if you intend to compare the images, setting a consistent color is also 
reasonable:

```bash
juliart generate --ca 0.5 --cb 0.5 --outfile ca-0.5-cb-0.5.png --rgb 90,12,10
```

For a bigger deep dive into this exercise (I was curious) see [Juliart Grid](https://github.com/vsoch/juliart-grid).


##### Text

As an alternative to file name (or just to have some fun!) you can generate
an image with a custom message.

```bash
juliart generate --text "Dinosaurs are great!"
```

By default the text will be printed in the top left corner (coordinate 10,10)
however you can adjust this to your liking:

```bash
$ juliart generate --text "Avocados are also great!" --xcoord 200 --ycoord 20
```

You can also change the font from OpenSans-Regular.ttf to Pacifico-Regular.ttf,
and choose an alpha (transparency).

```bash
$ juliart generate --text "Avocados!" --fontsize 30 --font-alpha 90 --font Pacifico-Regular
```

You can see all of these examples in the [text](img/text) examples folder.


##### Zoom, Iterations, and Radius

Otherwise you can do any of the customizations shown above! Try playing
around with iterations, colors/themes, and zoom to see different effects.

```bash
$ juliart generate --zoom 3
$ juliart generate --iter 100
```

After looking over the equation to generate the JuliaSet for a while,
I got the insight that if we increase the iterations value and modify the radious,
 this will generate more detailed images. Here are some examples:

```bash
juliart generate --iter 1000 --radius 8
```

![img/random/frigid-kitty-0005.png](https://raw.githubusercontent.com/vsoch/juliart/master/img/random/frigid-kitty-0005.png)

```bash
juliart generate --iter 5000 --radius 10
```

This one takes longer, of course.

##### Interactive Python

To generate from within Python, here is a quick example:

```python
from juliart.main import JuliaSet

juliaset = JuliaSet(
    resolution=1000, # (1000x1000)
    color="random", # random, glow, pattern
    theme="random",
    rgb=None,  # "197,18,12"
    iterations=200,
)
juliaset.generate_image(zoom=1.8)
juliaset.save_image("/tmp/myimage.png")
```

If you want to add text, before the save_image do:

```python
juliaset.write_text(
    text,
    fontsize=16,
    xcoord=10,
    ycoord=10,
)
```

By default, the text is white. You can change the color (and add a transparency, alpha layer)
like:


```python
juliaset.write_text(
    text,
    fontsize=16,
    xcoord=10,
    ycoord=10,
    rgb=(255, 255, 255, 15),
)
```

#### Animate

![img/animate/butterscotch-plant-7505.gif](https://raw.githubusercontent.com/vsoch/juliart/master/img/animate/butterscotch-plant-7505.gif)

You can inspect the parameters for the animation command as follows:

```bash
$ juliart animate --help
usage: juliart animate [-h] [--no-cleanup] [--constant-a] [--constant-b]
                       [--randomize-zoom] [--zoom-max ZOOM_MAX]
                       [--zoom-min ZOOM_MIN] [--frames FRAMES]
                       [--outfile OUTFILE] [--text TEXT] [--fontsize FONTSIZE]
                       [--xcoord XCOORD] [--ycoord YCOORD] [--ca CA] [--cb CB]
                       [--res RES] [--iter ITERS]
                       [--color {random,pattern,glow}] [--rgb RGB]
                       [--theme {christmas,easter,fall,random,halloween,hanukkah,spring,summer,thanksgiving,valentine,winter}]
                       [--zoom ZOOM]

optional arguments:
  -h, --help            show this help message and exit
  --no-cleanup          Do not delete temporary directory with png files to
                        generate gif.
  --constant-a          Don't randomize the point A on the circle.
  --constant-b          Don't randomize the point B on the circle.
  --randomize-zoom      Randomize the zoom up to --zoom-min or --zoom-max.
  --zoom-max ZOOM_MAX   the max zoom (must be greater than 3)
  --zoom-min ZOOM_MIN   the max zoom (must be greater than 0)
  --frames FRAMES       the number of frames to generate (default is 30)
  --outfile OUTFILE     the output file to save the image (defaults to
                        randomly generated png)
  --text TEXT           write a string of text to bottom left corner
  --fontsize FONTSIZE   font size of text (if desired) defaults to 16
  --xcoord XCOORD       x coordinate for text (defaults to 0)
  --ycoord YCOORD       y coordinate for text (defaults to 0)
  --ca CA               the a component of the c parameter
  --cb CB               the b component of the c parameter
  --res RES             the resolution to generate (defaults to 1000)
  --iter ITERS          the number of iterations per pixel (defaults to 200)
  --color {random,pattern,glow}
                        a color pattern to follow.
  --rgb RGB             a specific rbg color, in format R,G,B
  --theme {christmas,easter,fall,random,halloween,hanukkah,spring,summer,thanksgiving,valentine,winter}
                        a theme to color the art (defaults to random colors)
  --zoom ZOOM           the level of zoom (defaults to 1.8)
```

Most of these defaults (mins and maxes) are fairly reasonable and you wouldn't
want to change them. By default, we will vary both the A and B constants that
are used to generate the Julia Set, but we won't vary the zoom. 

##### Defaults

The defaults will generate a 30 framed animation, varying the parameters
A and B but not the zoom, with resolution (1000 X 1000) and 200 iterations.

```bash
$ juliart animate --frames 5
```

##### Optimize 

You can make that faster by changing the iterations and resolution, of course!

```bash
$ juliart animate --res 500 --iters 100
```

The tradeoff is image quality. You'll also notice in the case of generating
black pixels it takes slightly longer than white pixels (since white is 
an absence of color).

##### Text

You can add a consistent message to all frames in an animation:

```bash
juliart animate --text "Dinosaurs are great!"
```

and customize the `--xcoord`, `--ycoord`, or `--fontsize` parameters as
desired. See the [text](img/text) examples folder for an example. If you want
to vary the text between frames, then you'll need to roll your own
loop to generate the frames and then combine into an animation. An example
is provided, the code at [Juliart Grid](https://github.com/vsoch/juliart-grid).


##### Cleanup

If you want to keep the temporary png images (the frames) you can do:

```bash
$ juliart animate --no-cleanup
```

##### Frames

If you want to customize the number of frames, you can do that too (we default to 30). Here
would be a smaller / quicker to generate image:

```bash
$ juliart animate --frames 5
```

##### Zoom

To also randomize the zoom, specify:

```bash
$ juliart animate --randomize-zoom --frames 5
```

##### Python

To generate from within Python, it's fairly straight forward:

```python
from juliart.main import JuliaSetAnimation

juliaset = JuliaSetAnimation(
    resolution=args.res,
    color=args.color,
    theme=args.theme,
    rgb=args.rgb,
    cleanup=not args.skip_cleanup,
    iterations=args.iters,
    zoom_max=args.zoom_max,
    zoom_min=args.zoom_min,
)

juliaset.generate_animation(
    zoom=args.zoom,
    outfile=args.outfile,
    frames=args.frames,
    randomize_a=not args.constant_a,
    randomize_b=not args.constant_b,
    randomize_zoom=args.randomize_zoom,
)
```


### Colors

The three choices for colors are random, pattern, or glow, or setting your
own RGB value.

#### Random

Random is the default, and the image at the top of the README here is generated using
this setting. Take a look at more [more random examples](https://github.com/vsoch/juliart/tree/master/img/random).

#### Pattern

Pattern doesn't use a gradual gradient, but instead returns a hard boundary between
a color (and black). 

![img/pattern/delicious-lizard-8995.png](https://raw.githubusercontent.com/vsoch/juliart/master/img/pattern/delicious-lizard-8995.png)

Take a look at more [more pattern examples](https://github.com/vsoch/juliart/tree/master/img/pattern) here.

### RGB

If you want complete control of the color, provide a comma separated list
of RGB numbers as follows:

```bash
$ juliart generate --rgb 9,35,155
```

Note that this will also work with the `--color pattern` flag.

#### Glow

Glow means a dark background (black) with a color gradient. Here is an example:

![img/glow/dinosaur-diablo-1189.png](https://raw.githubusercontent.com/vsoch/juliart/master/img/glow/dinosaur-diablo-1189.png)

And is generated as follows:

```bash
juliart generate --color glow
```

See [more glow examples](https://github.com/vsoch/juliart/tree/master/img/glow).

If you choose glow, this will overwrite the choice of a theme (discussed next).

### Themes

To get a little more variety in your choice of colors, you can select a theme! 

```bash
juliart generate --theme halloween
juliart generate --theme christmas
juliart generate --theme hanukkah
juliart generate --theme thanksgiving
juliart generate --theme valentine

juliart generate --theme easter
juliart generate --theme fall
juliart generate --theme spring
juliart generate --theme summer
juliart generate --theme winter
```

For any of the commands above, you can also add `--color pattern` to flip the background
from the theme color to be black. 

```bash
$ juliart generate --theme halloween --color pattern
```

Take a look at images in the [themes folder](https://github.com/vsoch/juliart/tree/master/img/pattern) 
to get a sense of the color palettes.

### Docker

You can run the pre-generated Docker container too! You'll need to bind a folder
on the host to save the image to.

```bash
$ mkdir data
$ docker run -it -v $PWD/data/:/data quay.io/vanessa/juliart generate --outfile /data/art.png
```

You can also build the image first if you like:

```bash
$ docker build -t quay.io/vanessa/juliart .
```

The images are provided on [quay.io/vanessa/juliart](https://quay.io/repository/vanessa/juliart).

## Support

Do you have a question? Or want to suggest a feature to make it better?
Please [open an issue!](https://www.github.com/vsoch/juliart)


CHANGELOG.md
# CHANGELOG

This is a manually generated log to track changes to the repository for each release. 
Each section should include general headers such as **Implemented enhancements** 
and **Merged pull requests**. 

The versions coincide with releases on pypi.

## [0.0.x](https://github.com/vsoch/juliart/tree/master) (0.0.x)
 - bug with namespace (0.0.16)
 - bug that default alpha should be 100 (not 1) (0.0.15)
 - automated builds to quay.io and word wrap for longer text (0.0.14)
 - adding control for ca and cb parameters (0.0.13)
 - JuliaSetAnimation class to generation animations (0.0.12)
 - adding color themes (0.0.11)
 - initial release with basic generation for random, glow, pattern colors (0.0.1)



juliart/utils.py
"""

Copyright (C) 2019-2020 Vanessa Sochat.

This Source Code Form is subject to the terms of the
Mozilla Public License, v. 2.0. If a copy of the MPL was not distributed
with this file, You can obtain one at http://mozilla.org/MPL/2.0/.

Modified from https://github.com/Visual-mov/Colorful-Julia (MIT License)

"""

import os
import sys

here = os.path.dirname(os.path.abspath(__file__))


def get_font(filename="OpenSans-Regular.ttf"):
    """Return the default font for writing on the images. A user could
       add additional fonts to this folder, if desired.
    """
    font_file = os.path.join(here, "fonts", filename)
    if not os.path.exists(font_file):
        sys.exit("Font %s does not exist." % font_file)
    return font_file


def check_restricted(value, min_range, max_range):
    """Ensure that we have a float between a min and max range.
    """
    try:
        value = float(value)
    except ValueError:
        sys.exit("Parameter %s must be a float." % value)

    if value < min_range or value > max_range:
        sys.exit("ca and cb must be in range (-1, 1)")


juliart/colors.py
"""

Copyright (C) 2019-2020 Vanessa Sochat.

This Source Code Form is subject to the terms of the
Mozilla Public License, v. 2.0. If a copy of the MPL was not distributed
with this file, You can obtain one at http://mozilla.org/MPL/2.0/.

Modified from https://github.com/Visual-mov/Colorful-Julia (MIT License)

"""

from random import randint, choice
import os
import sys

christmas_colors = [
    (179, 0, 12),
    (220, 61, 42),
    (13, 239, 66),
    (0, 179, 44),
    (13, 89, 1),
]
hanukkah_colors = [
    (9, 35, 155),
    (154, 184, 194),
    (242, 235, 219),
    (242, 205, 60),
    (218, 174, 75),
]
valentine_colors = [
    (94, 8, 30),
    (181, 26, 58),
    (226, 71, 103),
    (228, 131, 151),
    (228, 205, 211),
]
halloween_colors = [
    (247, 95, 28),
    (255, 154, 0),
    (0, 0, 0),
    (136, 30, 228),
    (133, 226, 31),
]
fall_colors = [(96, 60, 20), (156, 39, 6), (212, 91, 18), (243, 188, 46), (95, 84, 38)]
summer_colors = [
    (35, 110, 150),
    (21, 178, 211),
    (255, 215, 0),
    (243, 135, 47),
    (255, 89, 143),
]
spring_colors = [
    (243, 168, 188),
    (245, 173, 148),
    (255, 241, 166),
    (180, 249, 165),
    (158, 231, 245),
]
winter_colors = [
    (66, 104, 124),
    (132, 165, 184),
    (179, 218, 241),
    (203, 203, 203),
    (112, 117, 113),
]
easter_colors = [
    (255, 212, 229),
    (224, 205, 255),
    (189, 232, 239),
    (183, 215, 132),
    (254, 255, 162),
]
thanksgiving_colors = [
    (108, 47, 0),
    (158, 104, 42),
    (241, 185, 48),
    (181, 71, 48),
    (138, 151, 72),
]


def get_theme_colors(theme):
    """Return a color from a themed set, the default being a random color
    """
    if theme == "christmas":
        return choice(christmas_colors)
    elif theme == "easter":
        return choice(easter_colors)
    elif theme == "fall":
        return choice(fall_colors)
    elif theme == "halloween":
        return choice(halloween_colors)
    elif theme == "hanukkah":
        return choice(hanukkah_colors)
    elif theme == "spring":
        return choice(spring_colors)
    elif theme == "summer":
        return choice(summer_colors)
    elif theme == "thanksgiving":
        return choice(thanksgiving_colors)
    elif theme == "valentine":
        return choice(valentine_colors)
    elif theme == "winter":
        return choice(winter_colors)
    else:
        return (randint(20, 200), randint(20, 200), randint(20, 200))


juliart/version.py
"""

Copyright (C) 2019-2020 Vanessa Sochat.

This Source Code Form is subject to the terms of the
Mozilla Public License, v. 2.0. If a copy of the MPL was not distributed
with this file, You can obtain one at http://mozilla.org/MPL/2.0/.

"""

__version__ = "0.0.16"
AUTHOR = "Vanessa Sochat"
AUTHOR_EMAIL = "vsochat@stanford.edu"
NAME = "juliart"
PACKAGE_URL = "http://www.github.com/vsoch/juliart"
KEYWORDS = "julia, julia set, art, holidays, generator, animation"
DESCRIPTION = "Command line tool for generating Julia Set graphics and animations"
LICENSE = "LICENSE"

################################################################################
# Requirements


INSTALL_REQUIRES = (("Pillow", {"min_version": "6.0.0"}),)
TESTS_REQUIRES = (("pytest", {"min_version": "4.6.2"}),)
ANIMATE_REQUIRES = (("imageio", {"min_version": "2.5.0"}),)

INSTALL_REQUIRES_ALL = INSTALL_REQUIRES + ANIMATE_REQUIRES


juliart/client.py
#!/usr/bin/env python

"""

Copyright (C) 2019-2020 Vanessa Sochat.

This Source Code Form is subject to the terms of the
Mozilla Public License, v. 2.0. If a copy of the MPL was not distributed
with this file, You can obtain one at http://mozilla.org/MPL/2.0/.

"""

from juliart.main import JuliaSet, JuliaSetAnimation
import juliart
import argparse
import sys
import os


def get_parser():
    parser = argparse.ArgumentParser(description="Julia Set art generator")
    parser.add_argument(
        "--version",
        dest="version",
        help="suppress additional output.",
        default=False,
        action="store_true",
    )

    description = "actions for Julia Set art generator"
    subparsers = parser.add_subparsers(
        help="juliart actions", title="actions", description=description, dest="command"
    )

    generate = subparsers.add_parser("generate", help="generate a Julia Set image")
    animate = subparsers.add_parser(
        "animate", help="create a Julia Set animation (gif)"
    )

    generate.add_argument(
        "--radius",
        dest="radius",
        help="the max radius to allow (default is 4)",
        type=int,
        default=4,
    )

    animate.add_argument(
        "--no-cleanup",
        dest="skip_cleanup",
        help="Do not delete temporary directory with png files to generate gif.",
        default=False,
        action="store_true",
    )

    animate.add_argument(
        "--constant-a",
        dest="constant_a",
        help="Don't randomize the point A on the circle.",
        default=False,
        action="store_true",
    )

    animate.add_argument(
        "--constant-b",
        dest="constant_b",
        help="Don't randomize the point B on the circle.",
        default=False,
        action="store_true",
    )

    animate.add_argument(
        "--randomize-zoom",
        dest="randomize_zoom",
        help="Randomize the zoom up to --zoom-min or --zoom-max.",
        default=False,
        action="store_true",
    )

    animate.add_argument(
        "--zoom-max",
        dest="zoom_max",
        help="the max zoom (must be greater than 3)",
        type=int,
        default=3,
    )

    animate.add_argument(
        "--zoom-min",
        dest="zoom_min",
        help="the max zoom (must be greater than 0)",
        type=int,
        default=0,
    )

    animate.add_argument(
        "--frames",
        dest="frames",
        help="the number of frames to generate (default is 30)",
        type=int,
        default=30,
    )

    for subparser in [generate, animate]:
        subparser.add_argument(
            "--outfile",
            dest="outfile",
            help="the output file to save the image (defaults to randomly generated png)",
            type=str,
            default=None,
        )

        ## Text

        subparser.add_argument(
            "--text",
            dest="text",
            help="write a string of text to bottom left corner",
            type=str,
            default=None,
        )

        subparser.add_argument(
            "--fontsize",
            dest="fontsize",
            help="font size of text (if desired) defaults to 16",
            type=int,
            default=16,
        )

        subparser.add_argument(
            "--font",
            dest="font",
            help="choice of font (defaults to open sans)",
            type=str,
            choices=["OpenSans-Regular", "Pacifico-Regular"],
            default="OpenSans-Regular",
        )

        subparser.add_argument(
            "--font-alpha",
            dest="font_alpha",
            help="font transparency (defaults to 1)",
            type=int,
            default=100,
        )

        subparser.add_argument(
            "--xcoord",
            dest="xcoord",
            help="x coordinate for text (defaults to 0)",
            type=int,
            default=10,
        )

        subparser.add_argument(
            "--ycoord",
            dest="ycoord",
            help="y coordinate for text (defaults to 0)",
            type=int,
            default=10,
        )

        subparser.add_argument(
            "--ca",
            dest="ca",
            help="the a component of the c parameter",
            type=float,
            default=None,
        )

        subparser.add_argument(
            "--cb",
            dest="cb",
            help="the b component of the c parameter",
            type=float,
            default=None,
        )

        subparser.add_argument(
            "--res",
            dest="res",
            help="the resolution to generate (defaults to 1000)",
            type=int,
            default=1000,
        )

        subparser.add_argument(
            "--iter",
            dest="iters",
            help="the number of iterations per pixel (defaults to 200)",
            type=int,
            default=200,
        )

        subparser.add_argument(
            "--color",
            dest="color",
            choices=["random", "pattern", "glow"],
            help="a color pattern to follow.",
            type=str,
            default="random",
        )

        subparser.add_argument(
            "--rgb",
            dest="rgb",
            help="a specific rbg color, in format R,G,B",
            type=str,
            default=None,
        )

        subparser.add_argument(
            "--theme",
            dest="theme",
            choices=[
                "christmas",
                "easter",
                "fall",
                "random",
                "halloween",
                "hanukkah",
                "spring",
                "summer",
                "thanksgiving",
                "valentine",
                "winter",
            ],
            help="a theme to color the art (defaults to random colors)",
            type=str,
            default="random",
        )

        subparser.add_argument(
            "--zoom",
            dest="zoom",
            help="the level of zoom (defaults to 1.8)",
            type=float,
            default=1.8,
        )

    return parser


def main():
    """main is the entrypoint to the juliart client.
    """

    parser = get_parser()

    # Will exit with subcommand help if doesn't parse
    args, extra = parser.parse_known_args()

    # Show the version and exit
    if args.version:
        print(juliart.__version__)
        sys.exit(0)

    # If the provided font doesn't end in ttf
    if hasattr(args, "font"):
        font = args.font
        if not args.font.endswith(".ttf"):
            font = "%s.ttf" % (font)

    # Initialize the JuliaSet
    if args.command == "generate":
        juliaset = JuliaSet(
            resolution=args.res,
            color=args.color,
            ca=args.ca,
            cb=args.cb,
            theme=args.theme,
            rgb=args.rgb,
            iterations=args.iters,
        )
        juliaset.generate_image(zoom=args.zoom, radius=args.radius)

        # Add text, if the user wants to (args.text will be checked to be None)
        juliaset.write_text(
            args.text,
            fontsize=args.fontsize,
            xcoord=args.xcoord,
            ycoord=args.ycoord,
            font=font,
            rgb=(255, 255, 255, args.font_alpha),
        )

        juliaset.save_image(args.outfile)

    elif args.command == "animate":

        # Takes similar args to JuliaSet, but also animation preferences
        juliaset = JuliaSetAnimation(
            resolution=args.res,
            color=args.color,
            theme=args.theme,
            rgb=args.rgb,
            ca=args.ca,
            cb=args.cb,
            cleanup=not args.skip_cleanup,
            iterations=args.iters,
            zoom_max=args.zoom_max,
            zoom_min=args.zoom_min,
        )

        juliaset.generate_animation(
            zoom=args.zoom,
            outfile=args.outfile,
            frames=args.frames,
            randomize_a=not args.constant_a,
            randomize_b=not args.constant_b,
            randomize_zoom=args.randomize_zoom,
            text=args.text,
            font=font,
            fontalpha=args.font_alpha,
            fontsize=args.fontsize,
            xcoord=args.xcoord,
            ycoord=args.ycoord,
        )

    else:
        parser.print_help()


if __name__ == "__main__":
    main()
